Decouple system timer and drivers

pizi-nordic opened the pull request:
During our work in timer area, we found that system clock frequency is used incorrectly in several drivers.
For example, it is used to obtain bus frequency needed to calculate baud rate of the UART interfaces.
If we select different system timer device, all these calculations become invalid.

This PR fixes this problem by obtaining all needed information directly from the DTS.

Brings us closer to: #15363

Note:
 This PR contains changes from https://github.com/zephyrproject-rtos/zephyr/pull/15619, https://github.com/zephyrproject-rtos/zephyr/pull/15514, https://github.com/zephyrproject-rtos/zephyr/pull/15505, https://github.com/zephyrproject-rtos/zephyr/pull/15424, https://github.com/zephyrproject-rtos/zephyr/pull/15415 and https://github.com/zephyrproject-rtos/zephyr/pull/15356 reworked according to @galak comments expressed on dev-meeting.

pizi-nordic committed:
drivers: uart_cmsdk_apb: Get clock frequency from DTS

The uart_cmsdk_apb driver used system clock frequency
as a base for baudrate calculation. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: k6x: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: uart_pl011: Get clock frequency from DTS

The uart_pl011 driver used system clock frequency
as a base for baudrate calculation. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: lpc54xxx: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: uart_cc32xx: Get clock frequency from DTS

The uart_cc32xx driver used system clock frequency
as a base for baudrate calculation. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: sam4s: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: kl2x: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
logging: log_backend_swo: Get clock frequency from DTS

The log_backend_swo used system clock frequency
as a base for SWO clock calculation. This commit
corrects that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: spi_dw: Get clock frequency from DTS

The spi_dw driver used system clock frequency
as a base for SPI bus frequency calculation.
This commit corrects that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: msp432p4xx: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: sam3x: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: uart_msp432p4xx: Get clock frequency from DTS

The uart_msp432p4xx driver used system clock frequency
as a base for baudrate calculation. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: samd2x: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: samde70: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: uart_stellaris: Get clock frequency from DTS

The uart_stellaris driver used system clock frequency
as a base for baudrate calculation. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: kwx: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: beetle_clock_control: Get CPU clock frequency from DTS

The clock control initialization code used system clock
frequency as a CPU clock frequency. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
arch: xtensa: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: snps_arc_iot: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: wdog_cmsdk_apb: Get clock frequency from DTS

The wdog_cmsdk_apb driver used system clock frequency
as a base for timeout calculation. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: uart_qmsi: Get clock frequency from DTS

The uart_qmsi driver used system clock frequency
as a base for baudrate calculation. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: i2c_cc32xx: Get clock frequency from DTS

The i2c_cc32xx driver used system clock frequency
as a base for I2C clock frequency calculation.
This commit corrects that by obtaining the needed value from DTS.

Please note, that for I2C devices the clock-frequency property
specifies SCK frequency, instead of frequency of the clock driving
peripheral. To solve that problem, a new property was added.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: pwm_qmsi: Do not use CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC

On some SoCs the frequency of the system clock is obtained at run time
as the exact configuration of the hardware is not known at compile time.
On such platforms using CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC define
directly introduces timing errors.

This commit replaces CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC by the call
to inline function sys_clock_hw_cycles_per_sec() which always returns
correct frequency of the system clock.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
debug: Do not use CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC

On some SoCs the frequency of the system clock is obtained at run time
as the exact configuration of the hardware is not known at compile time.
On such platforms using CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC define
directly introduces timing errors.

This commit replaces CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC by the call
to inline function sys_clock_hw_cycles_per_sec() which always returns
correct frequency of the system clock.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
logging: Do not use CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC

On some SoCs the frequency of the system clock is obtained at run time
as the exact configuration of the hardware is not known at compile time.
On such platforms using CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC define
directly introduces timing errors.

This commit replaces CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC by the call
to inline function sys_clock_hw_cycles_per_sec() which always returns
correct frequency of the system clock.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: timer: Do not use CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC

On some SoCs the frequency of the system clock is obtained at run time
as the exact configuration of the hardware is not known at compile time.
On such platforms using CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC define
directly introduces timing errors.

This commit replaces CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC by the call
to inline function sys_clock_hw_cycles_per_sec() which always returns
correct frequency of the system clock.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: i2c_bitbang: Do not use CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC

On some SoCs the frequency of the system clock is obtained at run time
as the exact configuration of the hardware is not known at compile time.
On such platforms using CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC define
directly introduces timing errors.

This commit replaces CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC by the call to
inline function sys_clock_hw_cycles_per_sec() which always returns
correct frequency of the system clock.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
tests: benchmarks: Do not use CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC

On some SoCs the frequency of the system clock is obtained at run time
as the exact configuration of the hardware is not known at compile time.
On such platforms using CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC define
directly introduces timing errors.

This commit replaces CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC by the call
to inline function sys_clock_hw_cycles_per_sec() which always returns
correct frequency of the system clock.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

anangl committed:
scripts: dts: Add extraction of clock frequency for fixed clock

This patch adds generation of `*_FIXED_CLOCK_FREQUENCY` macros for
clock consuming nodes that are provided with a fixed rate clock.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

pizi-nordic committed:
dts: bindings: Add fixed-clock DTS binding

This commit adds binding for clock provider with fixed frequency.

Signed-off-by: Derek Hageman <hageman@inthat.cloud>
Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
drivers: uart_cc13xx_cc26xx: Get clock frequency from DTS

The uart_cc13xx_cc26xx driver used system clock frequency
as a base for baudrate calculation. This commit corrects
that by obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: nxp_rt: Get system clock frequency from DTS

The SoC initialization code used timer clock frequency
as a system clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic committed:
soc: samr21: Get CPU clock frequency from DTS

The SoC initialization code used system clock frequency
as a CPU clock frequency. This commit corrects that by
obtaining the needed value from DTS.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

pizi-nordic commented:
@galak: As you requested these changes, I expect that you will be the first reviewer of this PR ;-)

pizi-nordic committed:
drivers: adc_stm32: Get clock frequency from DTS

The adc_stm32 driver used system timer frequency as a base for
busy-wait delay calculation. This commit corrects that by obtaining
the needed value from SystemCoreClock variable.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

mnkp commented on dts/arm/atmel/sam3x.dtsi:
```diff
@@ -15,6 +15,7 @@
 		cpu@0 {
 			device_type = "cpu";
 			compatible = "arm,cortex-m4";
+			clock-frequency = <84000000>;
```
I believe we shouldn't define clock frequency in the SoC dtsi file. This is board and application specific. Even as a default value. We don't do it for other parameters. The `clock-frequency` should be defined in the board .dts file.

pizi-nordic replied:
For boards based on the same SoC and having different clock frequency, the frequency is specified in board file. Otherwise I put the defaults in the SoC file.

If you think that this should be moved, free free to do so.

mnkp replied:
I understand that this PR was a lot of work and what I'm asking for is still more. However, until now we didn't put any board specific configuration parameters in the SoC .dtsi files. Your PR breaks this policy.

The possible clock frequency that the CPU can run with depends strictly on the quartz mounted on the PCB. This is board specific. Even SoCs with PLL, like this one, are limited by the choice of quartz. There is really no default clock frequency value, it either corresponds to the PCB or it is wrong. We can't define CPU frequency here just because some evaluation board happens to be shipped with some specific quartz and not the other.

What we could put in SoC .dtsi files is CPU frequency that can be configured with the built in RC oscillator. So any board could boot safely even if the board .dts file didn't provide clock-frequency property. But we still would need to touch all the .dts files of boards that use quartz and define clock-frequency property there.

Since we want to add clock-frequency property using .yaml files in a separate PR. I can help you by updating all the device drivers, SoC and board files for SiLabs and Atmel. When the other PR is merged.

pizi-nordic replied:
@mnkp: I did what you requested. Please re-review.

mnkp commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -426,9 +426,31 @@ def yaml_inc_error(msg):
     raise yaml.constructor.ConstructorError(None, None, msg)
 
 
+def extract_cpu_clock_frequencies():
```
A question: why don't we extract `clock-frequency` property via a .yaml binding file?

pizi-nordic replied:
The yaml file is parsed only if there is a "compatible" property, which is omitted in this node in most of the files. Also, such approach would require plenty of yaml files as we have plenty of different CPUs supported in Zephyr.

galak replied:
We should add the YAML files for the CPUs.  There can't be that many.

pizi-nordic replied:
We have 8 architectures in Zephyr. Only on ARM we have Cortex-M0, M0+, M3, M4, M4F, M7 and M33F. If we assume that we have 6 different CPUs in given architecture in average, this will require 48 different YAML files...

galak replied:
> We have 8 architectures in Zephyr. Only on ARM we have Cortex-M0, M0+, M3, M4, M4F, M7, M23 and M33F. If we assume that we have 6 different CPUs in given architecture in average, this will require 48 almost identical YAML files...

Then we have that many YAML files.  We can use the 'inherit' to reduce the duplication between them.

pizi-nordic replied:
Just for clarification. There will be dozens of files which will look like this:

```
---
title: ARM Cortex-M0 CPU
version: 0.1

description: ARM Cortex-M0 CPU binding

inherits:
    !include cpu-base.yaml

properties:
    compatible:
      constraint: "arm,cortex-m0"
...
```

```
---
title: ARM Cortex-M0+ CPU
version: 0.1

description: ARM Cortex-M0+ CPU binding

inherits:
    !include cpu-base.yaml

properties:
    compatible:
      constraint: "arm,cortex-m0+"
...
```

```
---
title: ARM Cortex-M3 CPU
version: 0.1

description: ARM Cortex-M3 CPU binding

inherits:
    !include cpu-base.yaml

properties:
    compatible:
      constraint: "arm,cortex-m3"
...
```

And this is what we want.
Am I correct?

galak replied:
Yes

pizi-nordic replied:
Discussed on slack. This is the approach we want.

pizi-nordic replied:
@galak: I have tried to go in the direction you suggested, but this really does no have much sense for me.

1. The number of YAML files is considerable.
2. We also have to introduce plenty of dts_fixup.h files, as defines generated have part of compatible string in the name.
3. For Cortex-M0+ we are getting `DT_ARM_CORTEX_M0+_0_CLOCK_FREQUENCY` define. So scripts modification is also needed to avoid `+` character (I assume that we will not change current compatible strings using `+` character).

Summarising: The effort required to handle this in the way you proposed is really big and mostly consists of creating work-arounds in dts_fixup.h files...

Do you really would like to go in this direction?

galak replied:
> @galak: I have tried to go in the direction you suggested, but this really does no have much sense for me.
> 
> 1. The number of YAML files is considerable.
> 2. We also have to introduce plenty of dts_fixup.h files, as defines generated have part of compatible string in the name.
> 3. For Cortex-M0+ we are getting `DT_ARM_CORTEX_M0+_0_CLOCK_FREQUENCY` define. So scripts modification is also needed to avoid `+` character (I assume that we will not change current compatible strings).
> 
> Summarising: The effort required to handle this in the way you proposed is really big and mostly consists of creating work-arounds in dts_fixup.h files...
> 
> Do you really would like to go in this direction?

Yes this is still the direction we should go.  I'd suggest we create a separate PR for the CPU node work.

galak replied:
> @galak: I have tried to go in the direction you suggested, but this really does no have much sense for me.
> 
> 1. The number of YAML files is considerable.
> 2. We also have to introduce plenty of dts_fixup.h files, as defines generated have part of compatible string in the name.
> 3. For Cortex-M0+ we are getting `DT_ARM_CORTEX_M0+_0_CLOCK_FREQUENCY` define. So scripts modification is also needed to avoid `+` character (I assume that we will not change current compatible strings).
> 
> Summarising: The effort required to handle this in the way you proposed is really big and mostly consists of creating work-arounds in dts_fixup.h files...
> 
> Do you really would like to go in this direction?

Yes this is still the direction we should go.  I'd suggest we create a separate PR for the CPU node work.

pizi-nordic replied:
Ok. How to handle the `+` in `DT_ARM_CORTEX_M0+_0_CLOCK_FREQUENCY` issue? Should we replace `+` with a `PLUS` in define name or change compatible string to something else?

pizi-nordic replied:
@galak: The work you asked for is done. The CPU frequency is extracted through YAML files and dts_fixup.h is used to gather back the defines where such step is needed.

Please check if this PR is now OK for you.

mnkp reviewed (changes requested):
@pizi-nordic Thanks, that makes things cleaner. I have some remarks though.

galak commented on dts/bindings/clock/fixed-clock.yaml:
```diff
@@ -0,0 +1,37 @@
+      generation: define
+      category: required
+
+    clocks:
```
Is this correct, seems like there shouldn't be a 'clocks' property here.

Sizurka replied:
This is used in #15585 for clocks that use another clock as a reference.  e.g. SAM0 DFLL, which has a fixed 48 MHz frequency but can run in closed loop mode multiplying another clock to 48 MHz.

galak replied:
Need to understand this more, but seems like this isn't a `fixed-clock` than.

galak commented on dts/bindings/clock/fixed-clock.yaml:
```diff
@@ -0,0 +1,37 @@
+      description: compatible strings
+      constraint: "fixed-clock"
+
+    label:
```
Do we need label?

Sizurka replied:
See #15585 again, the label is used to construct the clock tree: it's used to compare the input clock source (`..._CLOCK_CONTROLLER`) with the labels of the other clocks to determine the correct HAL constants to use to select the input clock.

galak replied:
Ok, will move the conversation of to #15585, but I think `label` isn't what we should be using.  Probably should be `clock-names` or `clock-output-names`

galak commented on dts/bindings/clock/fixed-clock.yaml:
```diff
@@ -0,0 +1,37 @@
+      type: array
+      category: optional
+      description: input clock source
+      generation: define
```
Add `#clock-cells` property here.

pizi-nordic replied:
Added.

galak commented on scripts/dts/extract/clocks.py:
```diff
@@ -153,6 +153,39 @@ def _extract_consumer(self, node_path, clocks, def_label):
+                if clock_provider['props']['compatible'] == 'fixed-clock':
+                    clock_prop_name = 'clock-frequency'
+                    clock_prop_label = 'FIXED_CLOCK_FREQUENCY'
+                    if clock_index == 0 and \
```
Is it possible for a 'fixed-clock' to be anything but a singleton?

pizi-nordic replied:
=> @anangl

anangl replied:
I'm not sure if I understood the question correctly but it seems perfectly possible that a device node references several clock nodes from which some are 'fixed-clocks' (for instance, one low- and one high-frequency clock, both with fixed rates). For such cases we need to add indexes to the generated labels.

galak commented on scripts/dts/extract/clocks.py:
```diff
@@ -153,6 +153,39 @@ def _extract_consumer(self, node_path, clocks, def_label):
+                # as a macro generated for the clock consumer.
+                if clock_provider['props']['compatible'] == 'fixed-clock':
+                    clock_prop_name = 'clock-frequency'
+                    clock_prop_label = 'FIXED_CLOCK_FREQUENCY'
```
want to change this to `CLOCKS_CLOCK_FREQUENCY`

pizi-nordic replied:
@anangl: What do you think about @galak suggestion?

anangl replied:
I'm fine with that.
I'm not sure if it makes it more clear what the label represents, but I don't have any better proposal.

galak replied:
I'm wanting this so its consistent with the property name.

pizi-nordic replied:
Changed to `CLOCKS_CLOCK_FREQUENCY`.

galak reviewed (changes requested):
See comments, also looks like a lot of 'bindings' have not been updated.

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -557,7 +557,7 @@ static int adc_stm32_init(struct device *dev)
 		LOG_ERR("ADC clock rate get error.");
 	}
 
-	wait_cycles = CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / adc_rate *
```
Use `SystemCoreClock` instead of `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`.
`SystemCoreClock` is updated after clock configuration, which won't happen on `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` or` DT_CPUS_CPU_0_CLOCK_FREQUENCY`

pizi-nordic replied:
Fixed.

erwango commented on dts/arm/st/f0/stm32f0.dtsi:
```diff
@@ -18,6 +18,7 @@
 		cpu@0 {
 			device_type = "cpu";
 			compatible = "arm,cortex-m0";
+			clock-frequency = <48000000>;
```
Clock frequency setting is done at board level as it depends on the clock generator selected/available on the board. 
So this should be removed from here.
Also, if `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` is not used on STM32, this is not needed and can be removed with no impact.

pizi-nordic replied:
Fixed.

pizi-nordic commented:
@erwango: Could you please check if this PR is OK for you now?

pizi-nordic committed:
dts: bindings: Add bindings for ARM, ARC, Xtensa CPUs

This commit adds YAML files providing bindings for the
ARM, ARC and Xtensa CPUs.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

anangl committed:
scripts: dts: Add special handling of '+' in 'str_to_label' function

This patch is needed in order to get compilable macros for compatibles
like "arm,cortex-m0+".

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

pizi-nordic commented:
@galak, @mnkp, @erwango: All your comments are applied (I hope :)). Please re-review.

pizi-nordic committed:
debug: tracing: Fix compilation error

This commit adds the kernel.h include in order to provide
sys_clock_hw_cycles_per_sec() declaration.

Signed-off-by: Piotr Zięcik <piotr.ziecik@nordicsemi.no>

mnkp reviewed (commented):
Looks good! I reviewed primarily modifications on Atmel and SiLabs SoCs and found only one minor issue. SiLabs series efr32fg1p, efr32mg12p are missing clock information. Following files need to be updated:
```
boards/arm/efr32mg_sltb004a/efr32mg_sltb004a.dts
boards/arm/efr32_slwstk6061a/efr32_slwstk6061a.dts
dts/arm/silabs/efr32mg.dtsi
dts/arm/silabs/efr32fg1p.dtsi
soc/arm/silabs_exx32/efr32mg12p/dts_fixup.h
soc/arm/silabs_exx32/efr32fg1p/dts_fixup.h
```
The clock information is not used by the drivers but required to configure SWO logger output.

ruuddw reviewed (commented):
Looks like I'm missing system clock definitions for the ARC nsim boards, for the rest looks ok to me.

galak reviewed (commented):
I'm trying to understand the usage here between the cpu clock-frequency, a system clock frequency and having full clock support for a given SoC.

galak commented:
> I'm trying to understand the usage here between the cpu clock-frequency, a system clock frequency and having full clock support for a given SoC.

I think we need to document (not sure where) that various sources of clock information and how they get used.

carlescufi commented:
@pizi-nordic I have split out this PR into 2 separate PRs:

https://github.com/zephyrproject-rtos/zephyr/pull/16957
and
https://github.com/zephyrproject-rtos/zephyr/pull/16958

pizi-nordic commented:
> Looks good! I reviewed primarily modifications on Atmel and SiLabs SoCs and found only one minor issue. SiLabs series efr32fg1p, efr32mg12p are missing clock information. Following files need to be updated:
> (...)

@mnkp: Updated. Please check. I will move changes to  #16957 and  #16958 soon.

mnkp reviewed (changes requested):
One SoC I've overlooked. The `same70` from Atmel also supports SWO logger output and needs to define `DT_CPU_CLOCK_FREQUENCY`.

mnkp commented:
A little bit late but I had a look at how clocks are defined in Linux DTS files. The "fixed-clock" DTS nodes are used primarily for the clocks generated directly by quartz or fixed by design at some predefined frequency. Thus "fixed-clock". They are not used for system clocks which - almost never - are fixed. Also, modern SoCs have clock trees so complex that it's not possible to talk about a single system clock any more. This is all captured by Linux DTS where clock trees are described using names corresponding to the data sheet and reflect their complexity.

This PR seems to be using "fixed-clock" compatible nodes as a mere convenience to extract the data. Maybe it would be better to skip defining all these "fixed-clock" nodes. Instead, everywhere where `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC` is used currently by the drivers we can replace it by `DT_CPU_CLOCK_FREQUENCY` define. And let the SoC maintainers take it from there to define proper DTS clock tree, if needed.

pizi-nordic commented:
@mnkp:

1. I have added missing fixup for samd70, however it is located in https://github.com/zephyrproject-rtos/zephyr/pull/16958

2. As agreed with @galak, my changes are not intended to abstract SoC clock tree. As I do not know all the hardware details, I only created black box(es) representing given clocks. For most of the SoCs, these clocks will have the same frequency as the CPU clock but this is not always the case - this is why bus clock is separated from CPU clock.

pizi-nordic commented:
Obsoleted by https://github.com/zephyrproject-rtos/zephyr/pull/16957 and https://github.com/zephyrproject-rtos/zephyr/pull/16958
